### PR TITLE
Revert "Revert "Fix corepack `packageManager` detection on monorepos""

### PR DESCRIPTION
**Pull request by EndangeredMassa**
Reverts vercel/vercel#11865

### CODE REVIEW

**Comment by onsclom** [comment]
Confirmed this is ready to merge! We were worried that this caused `ERR_INVALID_THIS`, but it turns out that was just another side effect of https://github.com/vercel/vercel/pull/11887#issuecomment-2299735222. [See more info](https://linear.app/vercel/issue/ZERO-2036#comment-8fa59c45).

File: packages/build-utils/test/unit.test.ts [hunk 1]
```diff
@@ -713,6 +713,43 @@ it('should detect package.json in nested frontend', async () => {
...
 });
 
+it('should detect `packageManager` in npm monorepo', async () => {
+  try {
+    process.env.ENABLE_EXPERIMENTAL_COREPACK = '1';
+
+    const base = path.join(__dirname, 'fixtures', '41-npm-workspaces-corepack');
+    const fixture = path.join(base, 'a');
+    const result = await scanParentDirs(fixture, false, base);
+    console.log(result);
```
Comment (EndangeredMassa) [inline packages/build-utils/test/unit.test.ts:723]: **suggestion (non-blocking):** ‏

```suggestion
```

File: packages/build-utils/test/unit.test.ts [hunk 2]
```diff
@@ -713,6 +713,43 @@ it('should detect package.json in nested frontend', async () => {
...
+  try {
+    process.env.ENABLE_EXPERIMENTAL_COREPACK = '1';
+    const base = path.join(
+      __dirname,
+      'fixtures',
+      '42-pnpm-workspaces-corepack'
+    );
+    const fixture = path.join(base, 'c');
+    const result = await scanParentDirs(fixture, false, base);
+    console.log(result);
```
Comment (EndangeredMassa) [inline packages/build-utils/test/unit.test.ts:743]: **suggestion (non-blocking):** ‏

```suggestion
```

File: packages/build-utils/src/fs/run-user-scripts.ts [hunk 3]
```diff
@@ -436,20 +446,35 @@ async function walkParentDirsMulti({
...
   for (const dir of traverseUpDirectories({ start, base })) {
     const fullPaths = filenames.map(f => path.join(dir, f));
     const existResults = await Promise.all(
       fullPaths.map(f => fs.pathExists(f))
     );
     const foundOneOrMore = existResults.some(b => b);
+    const packageJsonPath = path.join(dir, 'package.json');
+    const packageJson: PackageJson | null = await fs
+      .readJSON(packageJsonPath)
+      .catch(() => null);
```
Comment (EndangeredMassa) [inline packages/build-utils/src/fs/run-user-scripts.ts:464]: **suggestion (non-blocking):** ‏debug log the error

We should wrap calls like this in a function that can debug log the error and still return `null` if it fails.

File: packages/build-utils/test/unit.test.ts (hunk 2 above)
Comment (EndangeredMassa) [inline packages/build-utils/test/unit.test.ts:743]: **suggestion (non-blocking):** ‏

```suggestion
```

File: packages/build-utils/test/unit.test.ts (hunk 1 above)
Comment (EndangeredMassa) [inline packages/build-utils/test/unit.test.ts:723]: **suggestion (non-blocking):** ‏

```suggestion
```

**Review by onsclom — APPROVED** [review]

**Review by trek — APPROVED** [review]

**Pull request merged**